go-controller/pkg/cluster/master.go

```diff
@@ -355,14 +344,45 @@ func (cluster *OvnClusterController) SetupMaster(masterNodeName string) error {
 	return nil
 }
 
-func (cluster *OvnClusterController) addNode(node *kapi.Node) error {
+// Wrapper used during host subnet allocation for cleaner release of the
+// allocated subnet on errors
+type hostSubnet struct {
+	allocator *netutils.SubnetAllocator
+	subnet    *net.IPNet
+	firstIP   string
+}
+
+func newHostSubnet(subnet string, allocator *netutils.SubnetAllocator) (*hostSubnet, error) {
+	ip, ipNet, err := net.ParseCIDR(subnet)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse subnet %s: %v", subnet, err)
+	}
+	ip = util.NextIP(ip)
+	n, _ := ipNet.Mask.Size()
+
+	return &hostSubnet{
+		subnet:    ipNet,
+		allocator: allocator,
+		firstIP:   fmt.Sprintf("%s/%d", ip.String(), n),
+	}, nil
+}
+
+func (hsn *hostSubnet) release() error {
```

Comment:
This is confusing - it seems like `allocator` will be nil every time we restart the process? When is this called?

Reply:
@squeed allocator will be nil of the node already has a hostsubnet allocated when the master starts. But that's fine, we don't need to release that since it's already been allocated.

allocator will be !nil when we add a new node that has no subnet yet. That's when we need to release it within the context of addNode(), but I didn't want to have to pass back both a subnet and the subnet allocator itself. We could do that though; it just seemed icky.

Reply:
@squeed re-pushed to remove the allocator thing. It's a bit less clean in addNode() but much less code overall.